app/src/ui/undo/warn-local-changes-before-undo.tsx

```diff
@@ -0,0 +1,112 @@
+import * as React from 'react'
+import { Dialog, DialogContent, DialogFooter } from '../dialog'
+import { Repository } from '../../models/repository'
+import { Dispatcher } from '../dispatcher'
+import { Row } from '../lib/row'
+import { OkCancelButtonGroup } from '../dialog/ok-cancel-button-group'
+import { Commit } from '../../models/commit'
+
+interface IWarnLocalChangesBeforeUndoProps {
+  readonly dispatcher: Dispatcher
+  readonly repository: Repository
+  readonly commit: Commit
+  readonly isWorkingDirectoryClean: boolean
+  readonly onDismissed: () => void
+}
+
+interface IWarnLocalChangesBeforeUndoState {
+  readonly isLoading: boolean
+}
+
+/**
+ * Dialog that alerts user that there are uncommitted changes in the working
+ * directory where they are gonna be undoing a commit.
+ */
+export class WarnLocalChangesBeforeUndo extends React.Component<
+  IWarnLocalChangesBeforeUndoProps,
+  IWarnLocalChangesBeforeUndoState
+> {
+  public constructor(props: IWarnLocalChangesBeforeUndoProps) {
+    super(props)
+    this.state = { isLoading: false }
+  }
+
+  public render() {
+    const title = __DARWIN__ ? 'Undo Commit' : 'Undo commit'
+
+    return (
+      <Dialog
+        id="warn-local-changes-before-undo"
+        type="warning"
+        title={title}
+        loading={this.state.isLoading}
+        disabled={this.state.isLoading}
+        onSubmit={this.onSubmit}
+        onDismissed={this.props.onDismissed}
+      >
+        <DialogContent>
+          <Row>{this.getWarningText()}</Row>
+        </DialogContent>
+        <DialogFooter>
+          <OkCancelButtonGroup destructive={true} okButtonText="Continue" />
+        </DialogFooter>
+      </Dialog>
+    )
+  }
+
+  private getWarningText() {
+    if (
+      this.props.commit.isMergeCommit &&
+      !this.props.isWorkingDirectoryClean
+    ) {
+      return (
+        <>
+          You have changes in progress. Undoing the merge commit might result in
+          some of these changes being lost.
+          <br />
+          <br />
+          Also, undoing a merge commit will apply the changes from the merge
```

Comment:
Nit: maybe we could put the large repeated part of the merge commit warning into a variable at the top so if it ever needs modified we don't have to double check this chunk of text twice.

Reply:
Fixed!